File: day01/src/main.rs
use std::cmp::Reverse;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let nums = {
        let mut nums = Vec::new();
        let mut current = 0;
        for line in reader.lines() {
            let line = line?;
            if line == "" {
                nums.push(current);
                current = 0;
                continue;
            }
            current += line.parse::<i32>()?;
        }
        nums.push(current);
        nums.sort_by_key(|&k| Reverse(k));
        nums
    };

    if let [a, ..] = nums[..] {
        println!("Part 1: {}", a);
    }
    if let [a, b, c, ..] = nums[..] {
        println!("Part 2: {}", a + b + c);
    }
    Ok(())
}

File: day01/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var weights []int
	current := 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			weights = append(weights, current)
			current = 0
			continue
		}

		num, err := strconv.Atoi(line)
		if err != nil {
			log.Fatalln(err)
		}
		current += num
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	weights = append(weights, current)
	sort.Ints(weights)
	if len(weights) == 0 {
		return
	}
	last := len(weights) - 1
	fmt.Println("Part 1:", weights[last])
	if last < 2 {
		return
	}
	fmt.Println("Part 2:", weights[last]+weights[last-1]+weights[last-2])
}
